sentry-apollo-3/src/main/java/io/sentry/apollo3/SentryApollo3HttpInterceptor.kt

```diff
@@ -0,0 +1,165 @@
+    HttpInterceptor {
+
+    constructor(hub: IHub) : this(hub, null)
+    constructor(beforeSpan: BeforeSpanCallback) : this(HubAdapter.getInstance(), beforeSpan)
+
+    override suspend fun intercept(
+        request: HttpRequest,
+        chain: HttpInterceptorChain
+    ): HttpResponse {
+        val activeSpan = hub.span
+        return if (activeSpan == null) {
+            chain.proceed(request)
+        } else {
+            val span = startChild(request, activeSpan)
+
+            val requestBuilder = request.newBuilder()
+
+            if (TracingOrigins.contain(hub.options.tracingOrigins, request.url)) {
+                val sentryTraceHeader = span.toSentryTrace()
+                val baggageHeader = span.toBaggageHeader()
+                requestBuilder.addHeader(sentryTraceHeader.name, sentryTraceHeader.value)
+
+                baggageHeader?.let {
+                    requestBuilder.addHeader(it.name, it.value)
+                }
+            }
+
+            val modifiedRequest = requestBuilder.build()
+
+            return try {
+                val httpResponse = chain.proceed(modifiedRequest)
+                span.status = SpanStatus.fromHttpStatusCode(httpResponse.statusCode, SpanStatus.UNKNOWN)
+                finish(span, modifiedRequest, httpResponse)
+
+                httpResponse
+            } catch (e: Throwable) {
+                when (e) {
+                    is ApolloNetworkException -> span.status = SpanStatus.INTERNAL_ERROR
+                    else -> SpanStatus.UNKNOWN
+                }
+                finish(span, modifiedRequest)
+                throw e
+            }
+        }
+    }
+
+    private fun Long?.ifHasValidLength(fn: (Long) -> Unit) {
+        if (this != null && this != -1L) {
+            fn.invoke(this)
+        }
+    }
+
+    private fun startChild(request: HttpRequest, activeSpan: ISpan): ISpan {
+        val buffer = Buffer()
+        request.body?.writeTo(buffer)
+
+        val reader = InputStreamReader(buffer.inputStream())
+
+        val content = try {
+            hub.options.serializer.deserialize(
+                reader,
+                ApolloRequestBodyContent::class.java,
+                ApolloRequestBodyContent.Deserializer
+            )
+        } catch (e: Throwable) {
+            hub.options.logger.log(SentryLevel.ERROR, "Error deserializing Apollo Request Body.", e)
+        }
+
+        val span = when (content) {
+            is ApolloRequestBodyContent -> createSpanFromBodyContent(activeSpan, request, content)
+            else -> activeSpan.startChild("apollo.request.unknown")
+        }
+
+        val operationId = request.headers.firstOrNull { it.name == "X-APOLLO-OPERATION-ID" }?.value ?: "unknown"
+
+        return span.apply {
+            setData("operationId", operationId)
+        }
+    }
+
+    private fun createSpanFromBodyContent(activeSpan: ISpan, request: HttpRequest, content: ApolloRequestBodyContent): ISpan {
```

Comment:
Is it worth deserializing the body for retrieving operation and variables? Should this be an opt-in?

Also is there a PII concern with variables?